lib/pack.js

```diff
@@ -45,22 +45,34 @@ class Pack extends BaseCommand {
       args = ['.']
 
     const unicode = this.npm.config.get('unicode')
+    const dryRun = this.npm.config.get('dry-run')
 
-    // clone the opts because pacote mutates it with resolved/integrity
-    const tarballs = await Promise.all(args.map(async (arg) => {
+    // Get the manifests and filenames first so we can bail early on manifest
+    // errors before making any tarballs
+    const manifests = []
+    for (const arg of args) {
       const spec = npa(arg)
-      const dryRun = this.npm.config.get('dry-run')
       const manifest = await pacote.manifest(spec, this.npm.flatOptions)
+      if (!manifest._id)
+        throw new Error('Invalid package, must have name and version')
```

Comment:
this might be effectively a breaking change, I can see people having pipelines running `npm pack` running in which they have a `package.json` with no name or version...

maybe a safer approach is to just log a warning message for now and fix the filename result? It might still break these theoretical pipelines but at least in a way which is fixing it so `pkg-undefined.tgz` becomes `pkg.tgz` and maybe use https://github.com/npm/name-from-folder if a name is not found?

Reply:
i agree; it seems better to do the thing pack is supposed to do (make a tarball) without worrying about a valid packument, which is only a problem for publishing, not packing.

Reply:
But `npm pack` does not just create a stream of tarball data, it creates a tarball _file_, named by the name and version in the package.json file.

npm 6 did crash with `npm ERR! package.json requires a "name" field`, so I think the lack of this error is an unintentionally breaking change.

Reply:
cool, thanks for the clarification and sorry I mislead you @ljharb 😅